### ISSUE TITLE
Figure4

### ISSUE THREAD

**Issue by Curiosity-Mars** [Curiosity-Mars/total-testosterone#9]
# ============================================================
# Re-validation pipeline for Figure 4 (Correlation network)
# - Data: /content/drive/MyDrive/CommmedpatientTT.xlsx
# - Subjects: complete age & BMI  -> N=5854
# - Remaining missing: mean imputation (column-wise)
# - Standardization: z-score
# - Clustering: KMeans(K=4, random_state=0, n_init=50)
# - Correlations: Spearman within each cluster (original scale, after cohort filter)
# - Network edges: |r| >= THRESH (default 0.2)
# - Node "centrality": degree (number of edges above threshold)
# - Outputs: CSV + PNG per cluster, plus a combined ZIP for download
# ============================================================

import os
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import zipfile

from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler
from sklearn.cluster import KMeans

import networkx as nx
from scipy.stats import spearmanr

# -------------------------
# Colab download helper
# -------------------------
try:
    from google.colab import files
    IN_COLAB = True
except Exception:
    IN_COLAB = False

# -------------------------
# 0) Paths
# -------------------------
XLSX_PATH = "/content/drive/MyDrive/CommmedpatientTT.xlsx"
assert os.path.exists(XLSX_PATH), f"File not found: {XLSX_PATH}"

OUTDIR = "/content/revalidation_fig4_outputs"
os.makedirs(OUTDIR, exist_ok=True)

# -------------------------
# 1) Load
# -------------------------
df0 = pd.read_excel(XLSX_PATH)
print("Loaded:", df0.shape)

# Column harmonization (just in case)
rename_map = {
    "HDL": "HDL-C",
    "LDL": "LDL-C",
    "HDL_C": "HDL-C",
    "LDL_C": "LDL-C",
    "BodyWeight": "BW",
    "Weight": "BW",
}
df0 = df0.rename(columns={k: v for k, v in rename_map.items() if k in df0.columns})

required_cols = [
    "age","height","BW","BMI","Hb","Hct","Alb","AST","ALT","ALP","Cre",
    "Glu","TG","HDL-C","LDL-C","LH","FSH","CRP","TT"
]
missing_cols = [c for c in required_cols if c not in df0.columns]
assert len(missing_cols) == 0, f"Missing required columns: {missing_cols}"

X_raw = df0[required_cols].copy()

# -------------------------
# 2) BMI reconstruction attempt (fill BMI only if missing and height/BW available)
#     Note: in your data, BMI missing remained 23 because height missing in those rows.
# -------------------------
h = pd.to_numeric(X_raw["height"], errors="coerce")
w = pd.to_numeric(X_raw["BW"], errors="coerce")

height_is_cm = h.median(skipna=True) > 3.0
h_m = h / 100.0 if height_is_cm else h
bmi_from_hw = w / (h_m ** 2)

n_bmi_missing_before = int(X_raw["BMI"].isna().sum())
mask_bmi_na = X_raw["BMI"].isna()
X_raw.loc[mask_bmi_na, "BMI"] = bmi_from_hw.loc[mask_bmi_na]
n_bmi_missing_after = int(X_raw["BMI"].isna().sum())

print("\n[BMI reconstruction]")
print("height unit assumed:", "cm" if height_is_cm else "m")
print("BMI missing before:", n_bmi_missing_before)
print("BMI missing after :", n_bmi_missing_after)

# -------------------------
# 3) Subjects filter: complete age & BMI
# -------------------------
mask_complete = X_raw["age"].notna() & X_raw["BMI"].notna()
X_subj = X_raw.loc[mask_complete].copy()
N = len(X_subj)
print("\n[Subjects]")
print("N subjects (age & BMI complete):", N)

# Missingness report (after cohort filter)
miss = X_subj.isna().sum().sort_values(ascending=False)
miss.to_csv(os.path.join(OUTDIR, "missingness_after_ageBMI_filter.csv"))
print("\nMissing values per column (top):")
print(miss.head(10))

# -------------------------
# 4) Mean imputation + Standardization for KMeans
# -------------------------
imputer = SimpleImputer(strategy="mean")
scaler = StandardScaler()

X_imp = imputer.fit_transform(X_subj.values)
X_std = scaler.fit_transform(X_imp)

# Safety check for finiteness (important for any downstream)
is_finite = np.isfinite(X_std).all()
print("\nAll finite after preprocessing:", is_finite)

# -------------------------
# 5) KMeans (final fixed setting)
# -------------------------
K = 4
RANDOM_STATE = 0
N_INIT = 50

kmeans = KMeans(n_clusters=K, random_state=RANDOM_STATE, n_init=N_INIT)
labels = kmeans.fit_predict(X_std)

sizes = pd.Series(labels).value_counts().sort_index()
print("\n[KMeans]")
print("K:", K, "random_state:", RANDOM_STATE, "n_init:", N_INIT)
print("Cluster sizes:", sizes.to_dict())

# Attach labels
df = X_subj.copy()
df["cluster"] = labels

# -------------------------
# 6) TT quartiles (overall cohort), define Q1 indicator if needed
# -------------------------
# Spearman uses continuous TT; if you want Q1 as a node, define binary indicator:
tt = pd.to_numeric(df["TT"], errors="coerce")
qcuts = tt.quantile([0.25, 0.50, 0.75]).values

# Quartile bins (matching your earlier re-check)
# Q1: [min, q25], Q2: (q25, q50], Q3: (q50, q75], Q4: (q75, max]
# (Exact min/max may differ; we also output actual min/max per bin)
bins = [-np.inf, qcuts[0], qcuts[1], qcuts[2], np.inf]
df["TT_Q"] = pd.cut(df["TT"], bins=bins, labels=["Q1","Q2","Q3","Q4"], include_lowest=True)

tt_ranges = df.groupby("TT_Q")["TT"].agg(["min","max"])
tt_ranges.to_csv(os.path.join(OUTDIR, "TT_quartile_ranges.csv"))
print("\nTT quartile ranges:")
print(tt_ranges)

# Optional node: Q1 indicator
df["Q1_lowTT"] = (df["TT_Q"] == "Q1").astype(float)

# -------------------------
# 7) Spearman correlation networks per cluster
# -------------------------
THRESH = 0.2  # <- your manuscript threshold |r| < 0.2 filtered out => keep |r|>=0.2
VARS_FOR_NETWORK = [
    "age","height","BW","BMI","Hb","Hct","Alb","AST","ALT","ALP","Cre",
    "Glu","TG","HDL-C","LDL-C","LH","FSH","CRP","TT",
    "Q1_lowTT"  # include as categorical indicator node (optional but explicit)
]

# Helper: compute spearman r and p for all pairs
def spearman_pairwise(dfx: pd.DataFrame, cols: list):
    cols = [c for c in cols if c in dfx.columns]
    n = len(cols)
    edges = []
    for i in range(n):
        for j in range(i+1, n):
            a = cols[i]; b = cols[j]
            x = pd.to_numeric(dfx[a], errors="coerce")
            y = pd.to_numeric(dfx[b], errors="coerce")
            m = x.notna() & y.notna()
            nn = int(m.sum())
            if nn < 10:
                continue
            r, p = spearmanr(x[m], y[m])
            if np.isnan(r):
                continue
            edges.append([a, b, nn, float(r), float(p)])
    return pd.DataFrame(edges, columns=["var1","var2","N_pair","r","p_value"])

all_outputs = []

for c in range(K):
    dfc = df[df["cluster"] == c].copy()
    out_edges_all = spearman_pairwise(dfc, VARS_FOR_NETWORK)
    out_edges_all.to_csv(os.path.join(OUTDIR, f"cluster{c}_spearman_allpairs.csv"), index=False)

    # Filter by threshold
    out_edges = out_edges_all.loc[out_edges_all["r"].abs() >= THRESH].copy()
    out_edges = out_edges.sort_values(by="r", key=lambda s: s.abs(), ascending=False)
    out_edges.to_csv(os.path.join(OUTDIR, f"cluster{c}_network_edges_absr_ge_{THRESH}.csv"), index=False)

    # Build graph
    G = nx.Graph()
    for _, row in out_edges.iterrows():
        G.add_edge(row["var1"], row["var2"], r=row["r"], p=row["p_value"], N=row["N_pair"])

    # Degree (connectivity) = "centrality" used in text
    deg = dict(G.degree())
    deg_df = pd.DataFrame({"variable": list(deg.keys()), "degree_edges": list(deg.values())})
    deg_df = deg_df.sort_values("degree_edges", ascending=False)
    deg_df.to_csv(os.path.join(OUTDIR, f"cluster{c}_node_degree.csv"), index=False)

    # Plot network (node size proportional to degree; no axis)
    plt.figure(figsize=(7.5, 7.5))
    if G.number_of_nodes() == 0:
        plt.text(0.5, 0.5, f"Cluster {c}: no edges at |r| >= {THRESH}", ha="center", va="center")
        plt.axis("off")
    else:
        # Layout choice to reduce overlap
        pos = nx.kamada_kawai_layout(G)

        # Node sizes
        degrees = np.array([deg.get(n, 0) for n in G.nodes()], dtype=float)
        node_sizes = 200 + 250 * degrees  # scalable but readable

        # Edge widths (by |r|)
        edge_widths = [1.0 + 4.0*abs(G[u][v]["r"]) for u, v in G.edges()]

        nx.draw_networkx_edges(G, pos, width=edge_widths, alpha=0.6)
        nx.draw_networkx_nodes(G, pos, node_size=node_sizes, alpha=0.9)
        nx.draw_networkx_labels(G, pos, font_size=9)

        plt.title(f"Cluster {c} correlation network (Spearman, |r| ≥ {THRESH})")

        # no axes for network
        plt.axis("off")

    fig_path = os.path.join(OUTDIR, f"Figure4_cluster{c}_network.png")
    plt.tight_layout()
    plt.savefig(fig_path, dpi=300)
    plt.close()

    all_outputs += [
        f"cluster{c}_spearman_allpairs.csv",
        f"cluster{c}_network_edges_absr_ge_{THRESH}.csv",
        f"cluster{c}_node_degree.csv",
        f"Figure4_cluster{c}_network.png",
    ]

# -------------------------
# 8) Write a short "methods metadata" CSV for reproducibility
# -------------------------
meta = pd.DataFrame([{
    "N_loaded": len(X_raw),
    "N_ageBMI_complete": N,
    "BMI_missing_before": n_bmi_missing_before,
    "BMI_missing_after_reconstruct": n_bmi_missing_after,
    "height_unit_assumed": "cm" if height_is_cm else "m",
    "imputation": "column-wise mean (after age & BMI complete filter)",
    "standardization": "z-score (StandardScaler)",
    "kmeans_K": K,
    "kmeans_random_state": RANDOM_STATE,
    "kmeans_n_init": N_INIT,
    "correlation": "Spearman",
    "network_threshold_abs_r": THRESH,
    "centrality_definition": "degree = number of edges with |r| >= threshold",
    "Q1_node": "Q1_lowTT (binary indicator: TT quartile == Q1)",
}])
meta_path = os.path.join(OUTDIR, "revalidation_metadata.csv")
meta.to_csv(meta_path, index=False)

# -------------------------
# 9) ZIP everything + download
# -------------------------
zip_path = os.path.join(OUTDIR, "revalidation_fig4_outputs.zip")
with zipfile.ZipFile(zip_path, "w", compression=zipfile.ZIP_DEFLATED) as zf:
    # Core outputs
    zf.write(meta_path, arcname="revalidation_metadata.csv")
    zf.write(os.path.join(OUTDIR, "missingness_after_ageBMI_filter.csv"), arcname="missingness_after_ageBMI_filter.csv")
    zf.write(os.path.join(OUTDIR, "TT_quartile_ranges.csv"), arcname="TT_quartile_ranges.csv")
    # Cluster files
    for f in all_outputs:
        zf.write(os.path.join(OUTDIR, f), arcname=f)

print("\nZIP created:", zip_path)
print("Outputs dir:", OUTDIR)

if IN_COLAB:
    files.download(zip_path)
else:
    print("Not in Colab: download manually from:", zip_path)
